temporian/implementation/numpy/data/event.py

```diff
@@ -25,81 +25,148 @@
-    def data(self) -> Dict[Tuple, List[NumpyFeature]]:
+    def data(self) -> Dict[Tuple, IndexData]:
         return self._data
 
     @property
-    def sampling(self) -> NumpySampling:
-        return self._sampling
+    def feature_names(self) -> List[str]:
+        return self._feature_names
 
     @property
-    def _first_index_features(self) -> List[NumpyFeature]:
-        if self.first_index_value() is None:
-            return []
-        return self.data[self.first_index_value()]
+    def index_names(self) -> List[str]:
+        return self._index_names
+
+    @property
+    def is_unix_timestamp(self) -> bool:
+        return self._is_unix_timestamp
+
+    @property
+    def _first_index_features(self) -> List[np.ndarray]:
+        return list(self._data.values())[0].features
 
     @property
     def dtypes(self) -> Dict[str, DType]:
         return {
-            feature.name: feature.dtype
-            for feature in self._first_index_features
+            feature_name: DTYPE_MAPPING[feature.dtype.type]
+            for feature_name, feature in zip(
+                self._feature_names, self._first_index_features
+            )
         }
 
-    @sampling.setter
-    def sampling(self, sampling: NumpySampling) -> None:
-        self._sampling = sampling
+    @property
+    def feature_count(self) -> int:
+        return len(self._feature_names)
 
-    def first_index_value(self) -> Optional[Tuple]:
-        if self.data is None or len(self.data) == 0:
-            return None
+    def copy(self) -> NumpyEvent:
+        return deepcopy(self)
 
-        return next(iter(self.data))
+    def iterindex(self) -> Iterable[Tuple[Tuple, IndexData]]:
+        yield from self.data.items()
 
-    def first_index_features(self) -> List[NumpyFeature]:
-        if self.first_index_value() is None:
-            return []
-        return self.data[self.first_index_value()]
+    def index_dtypes(self) -> Dict[str, DType]:
+        return (
+            {
+                index_name: PYTHON_DTYPE_MAPPING[type(index_key)]
+                for index_name, index_key in zip(
+                    self._index_names, self.first_index_key()
+                )
+            }
+            if self._data
+            else {}
+        )
 
-    def feature_count(self) -> int:
-        return len(self.first_index_features())
+    def first_index_key(self) -> Optional[Tuple]:
+        if self._data is None or len(self._data) == 0:
+            return None
 
-    def feature_names(self) -> List[str]:
-        # Only look at the feature in the first index
-        # to get the feature names. All features in all
-        # indexes should have the same names
-        return [feature.name for feature in self.first_index_features()]
+        return next(iter(self._data))
+
+    def first_index_features(self) -> IndexData:
+        if self.first_index_key() is None:
+            return []
+        return self[self.first_index_key()]
 
     def schema(self) -> Event:
```

Comment:
The schema should not be recomputed from partial data. Instead, let's (1) keep a reference to the existing schema or (2) to pass both the data and schema to all the functions that require both.